cmake: modernize CMake infrastructure; adds modules, bindings; build Windows wheels

gsmecher opened the pull request:
This PR builds on the previous one (#20), so while it's a big change, it's not as sprawling as it seems.

My primary motivation is getting Windows wheels built for pygetdata. This created a fair cascade of tasks because:
- autotools only works on UNIX-like environments (including MSYS2/cygwin), but
- on Windows, native Python .pyd modules are really only supported using MSVC
- If we do that, the MacOS and Linux wheel builders should come along for the ride.

Practically speaking, native Windows support needs CMake instead of autotools, and switching the MacOS/Linux wheel builders at the same time makes sense. The autotools flow remains in place (and the github regression tests still pass against it.)

This PR contains a mix of "omnibus commits" (an incremental modernization of the CMake infrastructure does not make sense) and "drive-by commits" for individual bug fixes or improvements.

Note that we have definitely left Python 2.x behind (Numpy 2.x never supported it anyways).

gsmecher committed:
cmake modernization and python bindings

gsmecher committed:
big_test.py (Python tests): Python 2.x -> 3.x exception update

gsmecher committed:
CMake: correctly wrangle tests with exit code 77 (autotools convention for "skip me")

gsmecher committed:
.gitignore: ignore CMakeCache.txt and CMakeFiles

gsmecher committed:
bindings/python: move pyproject.toml to CMake backend

We now have two Python packaging/backend combinations:

- setup.py: setuptools, invoked explicitly by the autotools build, and
- pyproject.toml: scikit-build-core, invoked by the cmake build

It's slightly awkward having these two coexist side-by-side, but the
two build systems already do that.

gsmecher committed:
msvc: don't define FILENAME_MAX if it already exists

gsmecher committed:
big_test.py (Python tests): use path.join for "expected results" with paths

gsmecher committed:
big_test.py (Python tests): gracefully handle builds without regex/pcre support

gsmecher committed:
big_test.py (Python tests): CheckEOS: escape input before using re.search()

This has no impact on the current codebase, but in a future commit we'll
be adding Windows support (which means backslashes, which need to be
escaped.)

gsmecher committed:
Github workflows: add CMake regression tests and Windows wheels

gsmecher committed:
test/c89*: place GD_C89_API in an #ifndef guard to squash redefinition warnings

Since cmake on Windows unconditionally defines GD_C89_API, Windows
otherwise gives us warnings like:

\path\to\alter_crecip89.c(22,9): warning C4005: 'GD_C89_API': macro redefinition

gsmecher committed:
test/version_5: fix #ifdef within macro expansion

This works within GCC but not within MSVC, and the c++ preprocessor
documentation (see 16.3.4 in N4594) seems to disavow this behaviour.

gsmecher committed:
Windows: MSVC: Poison GD_TRUNC for now - it is unsafe to unlink while iterating

Without this poison, the following tests segfault:

	1426 - test_trunc (SEGFAULT)
	1427 - test_trunc_dir (SEGFAULT)
	1429 - test_trunc_truncsub (SEGFAULT)

gsmecher committed:
Remove HAVE_CONFIG_H macro entirely

Now that CMake generates gd_config.h (consistent with the automake
flow), this macro was redundant and can be safely removed.

gsmecher committed:
bindings/python: callback.py tests: win32 portability (CRLF and path separators)

gsmecher commented:
@ketiltrout and @arahlin, sorry to drop this on you before the holidays. This is a follow-up to #20 (and includes both series of commits).

gsmecher committed:
.github/workflows/package.yml: upgrade cibuildwheel; add Python 3.13/3.14

gsmecher committed:
bindings/python/pyproject.toml: Teach scikit-build to extract the version number

gsmecher committed:
package.yml: wheel builder: no need to SKIP Python 3.6/3.7 any more

Squashes the warning:

	Warning: cibuildwheel: Invalid skip selector: 'cp36*'.
	cibuildwheel 3.x no longer supports Python < 3.8. Please use the
	2.x series or update `skip`. This selector will have no effect.

	Warning: cibuildwheel: Invalid skip selector: 'cp37*'.
	cibuildwheel 3.x no longer supports Python < 3.8. Please use the
	2.x series or update `skip`. This selector will have no effect.

arahlin commented on bindings/cxx/test/big_test.cpp:
```diff
@@ -18,9 +18,7 @@
 // along with GetData; if not, write to the Free Software Foundation, Inc.,
 // 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 //
-#ifdef HAVE_CONFIG_H
```
Can you just add a `add_definitions()` call to just set HAVE_CONFIG_H, rather than editing all of these files to remove the ifdef?

gsmecher replied:
I believe HAVE_CONFIG_H is never actually unset any more - it dates back to when "gd_config.h" was actually called "config.h" (b974090), and I think it's obsolete (it's always #defined in all build scenarios - autotools and cmake.) It's removed here as a clean-up. I can move to `add_definitions()` but I don't think this boilerplate is useful any more.

arahlin replied:
I see, I was under the impression that there was still a possible build path that don't include this file.  This is fine as is then.

arahlin commented on bindings/python/setup.py.in:
```diff
@@ -95,7 +95,6 @@ setup(
         include_dirs = includes,
         library_dirs = [ join(libbuild_dir, '.libs') ],
         libraries = [ 'getdata' ],
-        define_macros = [ ('HAVE_CONFIG_H','1') ]
```
Is this source file not necessary anymore with the scikit build?

gsmecher replied:
There are two gd_config.h files implicated in the HAVE_CONFIG_H macro (and both are still necessary) - autotools generates from src/gd_config.h.in and CMake generates from cmake/gd_config.h.in. In both cases, the build ends up with (and needs) a gd_config.h.

arahlin replied:
Ah this was ambiguous -- I meant the setup.py.in file here, not the config file.

arahlin replied:
I guess the autotools build still uses setup.py?

gsmecher replied:
Ah - I understand. You're correct, autotools is tied to setup.py and CMake is tied to pyproject.toml. This is odd but I don't have a clear proposal for how to make it less odd (especially without further inflating the scope of this PR).

arahlin commented on cmake/CMakeLists.txt:
```diff
@@ -185,6 +383,9 @@ if (CMAKE_INSTALL_PREFIX_INITIALIZED_TO_DEFAULT)
 endif()
 
+# Generate gd_config.h
+configure_file(${CMAKE_SOURCE_DIR}/gd_config.h.in ${GD_DIR}/src/gd_config.h @ONLY)
```
```suggestion
configure_file(${CMAKE_SOURCE_DIR}/gd_config.h.in ${GD_DIR}/src/gd_config.h @ONLY)
add_definitions(-DHAVE_CONFIG_H)
```

gsmecher replied:
As above - I intended to get rid of HAVE_CONFIG_H, but for fairly soft reasons. If you feel it still belongs in place, that's totally fine and I'm happy to reinstate it.

arahlin commented on bindings/python/pyproject.toml:
```diff
@@ -1,16 +1,23 @@
 dynamic = ["version"]
 
-[tool.setuptools]
-script-files = ["../../util/dirfile2ascii", "../../util/checkdirfile"]
```
These utilities are actually quite useful to include in the wheels if possible.

gsmecher replied:
OK, halfway there. The wheels now

- produce a python module that's dynamically linked,
- include the corresponding .so/.dll, and
- include whichever of these two utilities can be compiled on the platform.

Unfortunately, two limitations

- dirfile2ascii uses getopt_long, which is unavailable on Windows, and
- pyproject.toml only allows Python scripts to be installed (via [project.scripts]), so for these binaries to be visible in PATH, we'd need to create Python wrappers for them.

arahlin replied:
I am losing all steam on making the windows build work, to be honest.  I'd just turn the utils off on the windows build altogether.  For the linux/osx builds, I think you can tell cmake to install binary files in the top level bin directory (something like `install(<target> DESTINATION bin)`) inside the wheel package, without any additional content in [project.scripts].  Does that work?

arahlin replied:
Perhaps `${SKBUILD_SCRIPTS_DIR}` is what you want? https://scikit-build-core.readthedocs.io/en/latest/guide/cmakelists.html#install-directories

arahlin replied:
Were you able to get these utilities included in the wheels?

gsmecher replied:
Scripts are now built and installed correctly (including appearing in $PATH) on hosts that support them (which notably excludes dirfile2ascii on Windows, because getopt_long isn't supported there.)

Note that I'm statically building everything in the wheels now - it turns out the dust has not settled on how to do the RPATH modifications for scripts that have shared-library dependencies:

* https://github.com/pypa/auditwheel/issues/340
* https://github.com/aestream/aestream/issues/112

Since that suggests statically linking the utilities, I've done the same for pygetdata.so as well (it would be the only other thing dynamically linking against libgetdata.so, so it's a wash from a size perspective.) Plugin shared libraries are still correctly audited/repaired and bundled into the .whl from the system.

Statically linking still feels like a horrifically bloated thing to do, but if the packaging system doesn't have clear, stable, and portable guidance, I'd rather eat the binary size than bump into either portability problems or churn obligations later on.

arahlin commented on cmake/gd_config.h.in:
```diff
@@ -0,0 +1,133 @@
+
+/* Package information */
+#define PACKAGE_NAME "GetData"
+#define PACKAGE_VERSION "0.11.0"
```
Should be able to pick this up from the PROJECT_VERSION cmake variable I think?

gsmecher replied:
Yup - see d626c9. This parses version strings from m4/version.m4 - which is pretty clunky, but functional.

arahlin commented on cmake/CMakeLists.txt:
```diff
@@ -1,13 +1,82 @@
-cmake_minimum_required(VERSION 2.6.4)
+cmake_minimum_required(VERSION 3.15)
 
 project(getdata)
```
Should include a version here.  Perhaps also some logic for whether this is run by scikit-build or manually.

gsmecher replied:
Similarly, d626c9. For the scikit-build vs. manual thing - what do you have in mind?

arahlin replied:
scikit-build-core uses setuptools-scm to get the version number.  I think you probably want to use that package name and version in the CMakeLists file if you're triggering it from the pyproject file, and otherwise fall back to the m4 parsing.  Perhaps something like this:

```
if (SKBUILD_PROJECT_NAME)
    project(${SKBUILD_PROJECT_NAME} VERSION ${SKBUILD_PROJECT_VERSION})
else()
    <parse m4/version.m4 here...>
    project(getdata VERSION ${GETDATA_VERSION})
endif()
```

arahlin replied:
This way you can use a single `PROJECT_VERSION` cmake variable everywhere else, which is set by the above project() call.

gsmecher replied:
Picking this up again - the current PR strictly parses version number from `m4/version.m4` for autotools/cmake but uses setuptools-scm in pyproject.toml. I would rather have a single source of truth. We are stuck with version.m4 (I don't think the autotools flow should move away from it), so parsing this instead of `setuptools-scm` might be better than using both in different build scenarios.

arahlin replied:
Is there a way to add a hook for setuptools-scm to pick up the version from the m4 file?

gsmecher replied:
Aha! Yes, I figured this out (it's not setuptools-scm, it's a regex parser baked into scikit-build-core.)

gsmecher committed:
CMake/wheels: properly configure and install utils (checkdirfile, dirfile2ascii)

gsmecher committed:
MSVC: work around #define snprintf _snprintf in stdio.h

gsmecher committed:
wheel build: use shared libraries; teach wheel repair steps how to find them

gsmecher committed:
CMake: Add BUILD_CXX compile option; disable it for pyproject.toml entry point

gsmecher committed:
CMake: Correct .dll placement for Windows

gsmecher committed:
CMake: extract GETDATA_VERSION macros from m4/version.m4

This is a little clunky but reduces the number of redundant places where
the version numbering must be updated.

arahlin commented on .github/workflows/package.yml:
```diff
@@ -87,43 +56,46 @@ jobs:
-            python -m pip install "numpy>=2.0" &&
-            .github/workflows/bootstrap.sh &&
-            make -C bindings/python pyconstants.c
-          CIBW_REPAIR_WHEEL_COMMAND_LINUX: >
-            LD_LIBRARY_PATH=$PWD/src/.libs auditwheel repair -w {dest_dir} {wheel}
-          CIBW_REPAIR_WHEEL_COMMAND_MACOS: >
-            DYLD_LIBRARY_PATH=$PWD/src/.libs delocate-wheel --require-archs {delocate_archs} -w {dest_dir} -v {wheel}
```
Are you sure these repair wheel command tweaks are not needed for linux and osx?  IIRC I've had to set the library paths by hand like this for the repair tools to find the shared libraries and set RPATH correctly.

arahlin commented on .github/workflows/package.yml:
```diff
@@ -87,43 +56,46 @@ jobs:
-            yum install -y bzip2-devel flac-devel xz-devel pcre-devel zziplib-devel &&
-            .github/workflows/bootstrap.sh --disable-python &&
-            make
+            yum install -y cmake bzip2-devel flac-devel xz-devel pcre-devel zziplib-devel
```
Are all of these plugin libraries being found correctly by cmake?  IIRC, I had to give hints to autotools with the bootstrap command to make sure that all of the. compression plugins built properly.

gsmecher replied:
I picked a recently built .whl file and tried it locally, and got

```
>>> pygetdata.encoding_support(pygetdata.AUTO_ENCODED)  # expected, this is not an encoding
-15
>>> pygetdata.encoding_support(pygetdata.BZIP2_ENCODED)  # GD_RDWR
1
>>> pygetdata.encoding_support(pygetdata.FLAC_ENCODED)  # GD_RDWR
1
>>> pygetdata.encoding_support(pygetdata.GZIP_ENCODED )  # GD_RDWR
1
>>> pygetdata.encoding_support(pygetdata.LZMA_ENCODED )  # GD_RDWR
1
>>> pygetdata.encoding_support(pygetdata.SIE_ENCODED )  # GD_RDWR
1
>>> pygetdata.encoding_support(pygetdata.SLIM_ENCODED )  # whoops, should be None
-14
>>> pygetdata.encoding_support(pygetdata.TEXT_ENCODED)  # GD_RDWR
1
>>> pygetdata.encoding_support(pygetdata.UNENCODED)  # GD_RDWR
1
>>> pygetdata.encoding_support(pygetdata.ZZIP_ENCODED)  # whoops - GD_RDONLY (should be 0)
None
>>> print(pygetdata.encoding_support(pygetdata.ZZSLIM_ENCODED)  # whoops, should be None
-14
```

Good news / bad news - this means the encodings (except for SLIM/ZZSLIM) are being correctly found and built into the .whl, but there's an underlying bug in the Python bindings. I have fixed the underlying bug in 2e0d49000.

Do you have a strong feeling about SLIM support?

arahlin replied:
SLIM hasn't been used for a long while and was always relatively niche, I think.  In a separate conversation with Steve / Don, we felt it was ok to drop it in the wheels.

arahlin commented on cmake/CMakeLists.txt:
```diff
@@ -23,37 +113,176 @@ macro(GD_FILES name folder)
+else()
+  math(EXPR GETDATA_INT_VERSION "${PROJECT_VERSION_MAJOR} * 10000 + ${PROJECT_VERSION_MINOR} * 100 + ${PROJECT_VERSION_PATCH}")
+endif()
+set(DEFINE_GD_GETDATA_INT_VERSION "#define GD_GETDATA_INT_VERSION ${GETDATA_INT_VERSION}")
```
This variable doesn't seem to be used anywhere?

gsmecher replied:
It's exported in getdata.h, so this is consistent between autotools and cmake. (I don't know if it's actually useful downstream.)

gsmecher committed:
bindings/python: fix encoding_support return codes

The DocStrings for the Python bindings report:

	encoding_support(encoding)

	The 'encoding' parameter should be one of the pygetdata.*_ENCODED
	symbols.  This method will return pygetdata.RDWR if the library can
	read and write the encoding, pygetdata.RDONLY if the library can
	only read the encoding, or None otherwise.  See
	gd_encoding_support(3).

However, the code returned None on GD_RDONLY==0 and the integer value of
the gd_encoding_support return code otherwise (typically GD_RDWR,
GD_E_UNKNOWN_ENCODING, or GD_E_UNSUPPORTED).

arahlin commented on .github/workflows/package.yml:
```diff
@@ -87,43 +56,46 @@ jobs:
-          CIBW_TEST_COMMAND: >
-            cd {package}/test &&
-            PYTHONPATH=`python -c 'import sysconfig; print(sysconfig.get_paths()["purelib"])'`:$PYTHONPATH make check
+          CIBW_TEST_COMMAND_LINUX: >
+            python -c "import pygetdata; print(f'pygetdata version: {pygetdata.__version__}')" &&
+            cd {project}/bindings/python/test &&
+            python big_test.py
+          CIBW_TEST_COMMAND_MACOS: >
+            python -c "import pygetdata; print(f'pygetdata version: {pygetdata.__version__}')" &&
+            cd {project}/bindings/python/test &&
+            python big_test.py
+          CIBW_TEST_COMMAND_WINDOWS: >
+            python -c "import pygetdata; print(f'pygetdata version: {pygetdata.__version__}')" &&
+            cd /d {project}\bindings\python\test &&
+            python big_test.py
```
Should this block just use the `ctest` utility to test everything?

gsmecher replied:
I think that would be redundant (moreover, it would balloon the build time - especially on windows). We already run the tests in the other GitHub actions. All I want to do here is run the Python kick-tests to ensure the import succeeds and the basic functionality there. It's testing that the CI/CD build worked, not that the library is free of bugs.

In fact, I think we should have a separate set of wheel-only tests that are pickier about what's compiled in. (We may want the wheel to insist on FLAC, for example, whereas the ctests should only verify against the pre-requisites that were present at compile time.)

gsmecher committed:
wheel build: teach arm64 macos to find, build, and install plugins

gsmecher committed:
wheel build: restructure .whl files; statically link

The static link, in particular, is intended to sidestep upstream issues
that still seem unsettled:

	https://github.com/pypa/auditwheel/issues/340
	https://github.com/aestream/aestream/issues/112

gsmecher committed:
pyproject.toml: parse version from m4/version.m4

arahlin commented on bindings/python/pyproject.toml:
```diff
@@ -1,16 +1,26 @@
+m4_define\(getdata_minor,\s*(?P<minor>\d+)\)[\s\S]*
+m4_define\(getdata_revision,\s*(?P<micro>\d+)\)
+'''
+result = "{major}.{minor}.{micro}"
```
Cool, learned something new!

arahlin commented on cmake/src/CMakeLists.txt:
```diff
@@ -35,11 +47,73 @@ endif()
+set_target_properties(getdata PROPERTIES
+  VERSION 11.0.0
+  SOVERSION 11
+  WINDOWS_EXPORT_ALL_SYMBOLS ON
```
Is this version separate from the package version?

arahlin reviewed (approved):
Looks great, I'm happy with how this is integrated while leaving the autotools build intact.  Thanks for all the hard work!

gsmecher committed:
CMake: correctly parse .so versioning information from m4/version.m4
